src/policy/rbf.cpp

```diff
@@ -174,3 +167,35 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
     }
     return std::nullopt;
 }
+
+std::optional<std::string> CheckAncestorScores(CAmount replacement_fees,
+                                               int64_t replacement_vsize,
+                                               const CTxMemPool::setEntries& ancestors,
+                                               const CTxMemPool::setEntries& all_conflicts)
+{
+    assert(!all_conflicts.empty());
+    // Ancestor score is the total modified fees divided by the total size.
+    // To get the ancestor score, add up all the individual modified fees and sizes. Don't try to
+    // use the cached ancestor fees and sizes, because entries may have overlapping ancestors.
+    for (CTxMemPool::txiter it : ancestors) {
+        replacement_fees += it->GetModifiedFee();
+        replacement_vsize += it->GetTxSize();
+    }
+    const CFeeRate replacement_ancestor_score(replacement_fees, replacement_vsize);
+    for (CTxMemPool::txiter it : all_conflicts) {
```

Comment:
If we want to fix the issue where a replacement transaction might have a worse mining score than some transactions it replaces, then I think the conservative thing to do would be to compare the ancestor score of the replacement transaction with the actual feerates of the conflicted transactions themselves (ie not the ancestor feerate of the conflicted transactions).

The rationale is that you never know if some sibling transaction might exist in the mempool that would cause the ancestors of a target transaction to be mined, allowing the target transaction to be considered using its own feerate (rather than being weighed down by low feerate parents). 

This may be overly conservative; I think the real question that we're trying to answer with the feerate checks is something like "are the fees expected to be paid in the next (N?) blocks higher or lower if we process this transaction", which is not very easy to answer with the tools we have at our disposal.  If we're okay being conservative here for the use cases that we're trying to support, then I think comparing ancestor score to actual feerates would be an easy fix though.